File: server/db/schemas.js
const {Sequelize, DataTypes} = require('sequelize');
require('dotenv').config();

// connect to postgres database instead
const sequelize = new Sequelize(process.env.POSTGRES_URI, {
  dialect: 'postgres',
  logging: false,
  pool: {
    max: 20,
    min: 5,
    acquire: 30000,
    idle: 10000
  }
});

const Like = sequelize.define('UserVideoLike', {
  like_value: {
    type: DataTypes.BOOLEAN,
    allowNull: true
  },
  user_id: {
    type: DataTypes.STRING,
    allowNull: false, // Ensure this field is required
  },
  video_id: {
    type: DataTypes.STRING,
    allowNull: false, // Ensure this field is required
  }
}, {
  indexes: [
    {
      name: 'user_video_id',
      fields: ['user_id', 'video_id'],
      unique: true
    }
  ]
});

const View = sequelize.define('UserVideoView', {
}, {
  // indexes: [
  //   {
  //     name: 'user_video_id',
  //     fields: ['user_id', 'video_id'],
  //     unique: true
  //   }
  // ]
});

const User = sequelize.define('Users', 
  {
    username: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true
    },
    email: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true
    },
    password: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    key: {
      type: DataTypes.STRING
    },
    userId: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true,
      primaryKey: true
    }
  },
  {
    indexes: [
      {
        name: 'users_email',
        unique: true,
        fields: ['email']
      },
      {
        name: 'users_username',
        unique: true,
        fields: ['username']
      }
    ]
  }
);

const Video = sequelize.define('Videos', 
  {
    id: {
      type: DataTypes.STRING,
      unique: true,
      allowNull: false,
      primaryKey: true
    },
    uploaded: {
      type: DataTypes.BOOLEAN,
      defaultValue: false
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false
    },
    author: {
      type: DataTypes.STRING,
      allowNull: false
    },
    likes: {
      type: DataTypes.INTEGER,
      defaultValue: 0
    },
    author_id: {
      type: DataTypes.STRING,
      references: {
        model: User,  // or 'Users' if not imported
        key: 'userId'
      },
      allowNull: false,  // Ensures it cannot be null
      onDelete: 'CASCADE',
      onUpdate: 'CASCADE'
    },
    views: {
      type: DataTypes.INTEGER,
      defaultValue: 0
    },
    description: {
      type: DataTypes.STRING,
      allowNull: false
    }
  }
)

module.exports = {
  sequelize, User, Video, Like, View
}

File: server/filtering/schemas.py
from sqlalchemy import create_engine, Column, String, Integer, Boolean, ForeignKey
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, sessionmaker

Base = declarative_base()

class User(Base):
    __tablename__ = 'Users'
    userId = Column(String, primary_key=True)
    username = Column(String, unique=True, nullable=False)
    email = Column(String, unique=True, nullable=False)
    password = Column(String, nullable=False)

    # videos = relationship('Videos', back_populates='author')
    # likes = relationship('UserVideoLike', back_populates='user_id')

class Video(Base):
    __tablename__ = 'Videos'
    id = Column(String, primary_key=True)
    title = Column(String, nullable=False)
    author_id = Column(String, ForeignKey('users.userId'), nullable=False)
    likes = Column(Integer, default=0)
    views = Column(Integer, default=0)
    description = Column(String, nullable=False)

    # author = relationship('User', back_populates='id')
    # video_likes = relationship('Like', back_populates='video_id')

class Like(Base):
    __tablename__ = 'UserVideoLikes'
    id = Column(Integer, primary_key=True)
    user_id = Column(String, ForeignKey('users.userId'))
    video_id = Column(String, ForeignKey('videos.id'))
    like_value = Column(Boolean, nullable=False)  # True for like, False for dislike

    # user = relationship('User', back_populates='userId')
    # video = relationship('Video', back_populates='id')